File: greedy_algorithm/763_Partition Labels.js
/**
 * @param {string} s
 * @return {number[]}
 */
var partitionLabels = function (s) {
    const res = [];
    const hash = new Array(26).fill(0);
    const base = 'a'.charCodeAt(0);
    for (let i = 0; i < s.length; i++) {//Collect the last position of every character
        hash[s.charCodeAt(i) - base] = i;
    }
    var start = 0;
    var end = 0;
    for (let i = 0; i < s.length; i++) {
        end = Math.max(hash[s.charCodeAt(i) - base], end);//Update the current furthest position
        if (i == end) {//If the string has reached to the furthest position, we can partition the string
            res.push(end - start + 1);
            start = i + 1;//Update the next starting position
        }
    }
    return res;
};

File: greedy_algorithm/763_Partition Labels.cpp
#include <iostream>
#include<string>
#include<vector>
using namespace std;
class Solution
{
public:
    vector<int> partitionLabels(string s)
    {
        int hash[27] = {0};
        vector<int> res;
        for (int i = 0; i < s.size(); i++)
        {
            hash[s[i] - 'a'] = i;
        }
        int left = 0;
        int right = 0;
        for (int i = 0; i < s.size(); i++)
        {
            right = max(right, hash[s[i] - 'a']);
            if (i == right)
            {
                res.push_back(right - left + 1);
                left = i + 1;
            }
        }
        return res;
    }
};